### PR TITLE
8.0: Parsing of parenthesis

### PR DESCRIPTION
**Pull request by nene**
Just the parenthesis parsing portion, which works with all current config options.

Managed to get all the tests working, except for one.

### CODE REVIEW

File: src/core/Parser.ts [hunk 1]
```diff
@@ -3,11 +3,30 @@ import { EOF_TOKEN, type Token, TokenType } from './token';
...
+
+export type Parenthesis = {
+  type: 'parenthesis';
+  children: AstNode[];
+  openParen: string;
+  closeParen: string;
+  hasWhitespaceBefore: boolean;
+};
+
+export type AstNode = Parenthesis | TokenNode;
```
Comment (inferrinizzard) [inline src/core/Parser.ts:23]: may want to export these types into a types def files, maybe src/types/parser.d.ts ?

File: src/core/StatementFormatter.ts [hunk 2]
```diff
@@ -19,34 +19,40 @@ export default class StatementFormatter {
...
   private currentNewline = true;
+  private inline = false;
   private previousReservedToken: Token = EOF_TOKEN;
   private previousCommandToken: Token = EOF_TOKEN;
-  private tokens: Token[] = [];
+  private nodes: AstNode[] = [];
   private index = -1;
 
-  constructor(cfg: FormatOptions, params: Params) {
+  constructor(cfg: FormatOptions, params: Params, { inline = false }: { inline?: boolean } = {}) {
```
Comment (inferrinizzard) [inline src/core/StatementFormatter.ts:28]: can we also extract flags param as a type

**Review by inferrinizzard — COMMENTED** [review]
a few nitpicks but looks good 👍

File: src/core/Parser.ts (hunk 1 above)
Comment (inferrinizzard) [reply src/core/Parser.ts:23]: addressed in future PR

File: src/core/StatementFormatter.ts (hunk 2 above)
Comment (nene) [reply src/core/StatementFormatter.ts:28]: Done in future PR.

File: src/core/StatementFormatter.ts [hunk 3]
```diff
@@ -116,47 +121,49 @@ export default class StatementFormatter {
...
-      if (type === TokenType.OPEN_PAREN) {
-        openBlocks++;
-      }
-      if (type === TokenType.CLOSE_PAREN) {
-        openBlocks--;
-      }
-    }
-    return count;
+  private countClauses(nodes: AstNode[]): number {
+    return 1 + nodes.filter(node => isTokenNode(node) && node.token.value === ',').length;
```
[the remark this answers is not in the document]
Comment (nene) [reply src/core/StatementFormatter.ts:154]: Deleted this code in future PR.

**Pull request merged**